lib/experimental/interactivity-api/directive-processing.php

```diff
@@ -0,0 +1,156 @@
+<?php
+/**
+ * Functions and hooks to process the server side rendering of the Interactivity
+ * API directives.
+ *
+ * @package Gutenberg
+ * @subpackage Interactivity API
+ */
+
+/**
+ * Process directives in each block.
+ *
+ * @param string $block_content The block content.
+ * @param array  $block         The full block.
+ *
+ * @return string Filtered block content.
+ */
+function gutenberg_interactivity_process_directives_in_root_blocks( $block_content, $block ) {
+	// Don't process inner blocks or root blocks that don't contain directives.
+	if ( isset( $block['is_inner_block'] ) || strpos( $block_content, 'data-wp-' ) === false ) {
+		return $block_content;
+	}
+
+	// TODO: Add some directive/components registration mechanism.
+	$directives = array(
+		'data-wp-bind'    => 'gutenberg_interactivity_process_wp_bind',
+		'data-wp-context' => 'gutenberg_interactivity_process_wp_context',
+		'data-wp-class'   => 'gutenberg_interactivity_process_wp_class',
+		'data-wp-style'   => 'gutenberg_interactivity_process_wp_style',
+		'data-wp-text'    => 'gutenberg_interactivity_process_wp_text',
+	);
+
+	$tags = new WP_Directive_Processor( $block_content );
+	$tags = gutenberg_interactivity_process_directives( $tags, 'data-wp-', $directives );
+	return $tags->get_updated_html();
+}
+add_filter( 'render_block', 'gutenberg_interactivity_process_directives_in_root_blocks', 10, 2 );
+
+/**
+ * Mark the inner blocks with a temporary property so we can discard them later,
+ * and process only the root blocks.
+ *
+ * @param array $parsed_block The parsed block.
+ * @param array $source_block The source block.
+ * @param array $parent_block The parent block.
+ *
+ * @return array The parsed block.
+ */
+function gutenberg_interactivity_mark_inner_blocks( $parsed_block, $source_block, $parent_block ) {
+	if ( isset( $parent_block ) ) {
+		$parsed_block['is_inner_block'] = true;
+	}
+	return $parsed_block;
+}
+add_filter( 'render_block_data', 'gutenberg_interactivity_mark_inner_blocks', 10, 3 );
+
+/**
+ * Process directives.
+ *
+ * @param WP_Directive_Processor $tags An instance of the WP_Directive_Processor.
+ * @param string                 $prefix Attribute prefix.
+ * @param string[]               $directives Directives.
+ *
+ * @return WP_Directive_Processor The modified instance of the
+ * WP_Directive_Processor.
+ */
+function gutenberg_interactivity_process_directives( $tags, $prefix, $directives ) {
+	$context   = new WP_Directive_Context;
+	$tag_stack = array();
+
+	while ( $tags->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+		$tag_name = strtolower( $tags->get_tag() );
```

Comment:
why are we lower-casing the tag name? I think someone explained this once, but is there any reason to do so?